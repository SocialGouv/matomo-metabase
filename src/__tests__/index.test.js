const mock_pgQuery = jest.fn();
const mock_matomoApi = jest.fn();

process.env.MATOMO_SITE = "42";
process.env.PROJECT_NAME = "some-project";

const matomoVisit = require("./visit.json");

const run = require("../index");

jest.mock("pg", () => {
  class Client {
    escapeIdentifier(name) {
      return name;
    }
    end() {}
    connect() {
      return Promise.resolve();
    }
    query(...args) {
      return mock_pgQuery(...args);
    }
  }
  return {
    Client,
  };
});

jest.mock("piwik-client", () => {
  class MatomoClient {
    api(...args) {
      return mock_matomoApi(...args);
    }
  }
  return MatomoClient;
});

beforeEach(() => {
  jest.resetAllMocks();
  jest.resetModules();
  process.env.STARTDATE = "";
});

test("run: should create table", async () => {
  mock_pgQuery.mockReturnValue({ rows: [{ action_timestamp: new Date("2020-01-01T00:00:00.000Z").getTime() }] });
  mock_matomoApi.mockImplementation((options, cb) => {
    return cb(null, []);
  });
  await run();
  expect(mock_pgQuery.mock.calls[0]).toMatchSnapshot();
});

test("run: should get the smart date base on the last event", async () => {
  jest.useFakeTimers("modern").setSystemTime(new Date("2021-08-01T00:00:00.000Z").getTime());
  mock_pgQuery.mockReturnValueOnce({ rows: [] });
  mock_pgQuery.mockReturnValue({ rows: [{ count: 0 }] });

  mock_matomoApi.mockImplementation((options, cb) => {
    return cb(null, [
      {
        ...matomoVisit,
        idVisit: 123,
      },
      {
        ...matomoVisit,
        idVisit: 124,
      },
    ]);
  });

  await run();

  const expectedDbCalls = [
    [
      "CREATE TABLE IF NOT EXISTS matomo\n" +
        "  (\n" +
        "    idsite                      text,\n" +
        "    idvisit                     text,\n" +
        "    actions                     text,\n" +
        "    country                     text,\n" +
        "    region                      text,\n" +
        "    city                        text,\n" +
        "    operatingsystemname         text,\n" +
        "    devicemodel                 text,\n" +
        "    devicebrand                 text,\n" +
        "    visitduration               text,\n" +
        "    dayssincefirstvisit         text,\n" +
        "    visitortype                 text,\n" +
        "    sitename                    text,\n" +
        "    userid                      text,\n" +
        "    serverdateprettyfirstaction date,\n" +
        "    action_id                   text UNIQUE,\n" +
        "    action_type                 text,\n" +
        "    action_eventcategory        text,\n" +
        "    action_eventaction          text,\n" +
        "    action_eventname            text,\n" +
        "    action_eventvalue           text,\n" +
        "    action_timespent            text,\n" +
        "    action_timestamp            timestamp with time zone,\n" +
        "    usercustomproperties        json\n" +
        ")",
      [],
    ],
    ["select * from matomo order by action_timestamp desc limit 1"],
  ];

  // check matomo requests
  expect(mock_matomoApi.mock.calls.length).toEqual(0);

  // check db requests
  expect(mock_pgQuery.mock.calls.length).toEqual(expectedDbCalls.length);

  // // check matomo requests
  // expect(mock_matomoApi.mock.calls.length).toEqual(expectedMatomoCalls.length);
  // expect(mock_matomoApi.mock.calls.map((c) => [c[0].date, c[0].idSite])).toEqual(expectedMatomoCalls);
  // // check the 4 events inserted
  // expect(mock_pgQuery.mock.calls.length).toEqual(1 + expectedMatomoCalls.length * (4 + 1));
  // expect(mock_pgQuery.mock.calls).toMatchSnapshot();
});

test("run: should use now if nothing in DB", async () => {
  jest.useFakeTimers("modern").setSystemTime(new Date("2020-01-01").getTime());
  mock_pgQuery.mockReturnValue({ rows: [] });

  mock_matomoApi.mockImplementation((options, cb) => {
    return cb(null, [
      {
        ...matomoVisit,
        idVisit: 123,
      },
    ]);
  });

  await run();

  expect(mock_matomoApi.mock.calls.length).toEqual(0);
  expect(mock_pgQuery.mock.calls.length).toEqual(2);


  // const expectedMatomoCalls = [
  //   ["2019-12-29", "42"],
  //   ["2019-12-30", "42"],
  //   ["2019-12-31", "42"],
  // ];

  // // check matomo requests
  // expect(mock_matomoApi.mock.calls.length).toEqual(expectedMatomoCalls.length);
  // expect(mock_matomoApi.mock.calls.map((c) => [c[0].date, c[0].idSite])).toEqual(expectedMatomoCalls);
  // // check the 4 events inserted
  // expect(mock_pgQuery.mock.calls.length).toEqual(1 + expectedMatomoCalls.length * (2 + 1));
  // expect(mock_pgQuery.mock.calls).toMatchSnapshot();
});

test("run: should use given date if any", async () => {
  jest.useFakeTimers("modern").setSystemTime(new Date("2020-01-01").getTime());
  mock_pgQuery.mockReturnValue({ rows: [] });

  mock_matomoApi.mockImplementation((options, cb) => {
    return cb(null, [
      {
        ...matomoVisit,
        idVisit: 123,
      },
    ]);
  });

  await run("2019-07-01");


  expect(mock_matomoApi.mock.calls.length).toEqual(184);

  // const expectedMatomoCalls = [
  //   ["2019-06-28", "42"],
  //   ["2019-06-29", "42"],
  //   ["2019-06-30", "42"],
  // ];

  // // check matomo requests
  // expect(mock_matomoApi.mock.calls.length).toEqual(expectedMatomoCalls.length);
  // expect(mock_matomoApi.mock.calls.map((c) => [c[0].date, c[0].idSite])).toEqual(expectedMatomoCalls);
  // // check the 4 events inserted
  // expect(mock_pgQuery.mock.calls.length).toEqual(1 + expectedMatomoCalls.length * (2 + 1));
  // expect(mock_pgQuery.mock.calls).toMatchSnapshot();
});

test("run: should use STARTDATE if any", async () => {
  process.env.STARTDATE = "2019-07-15";
  jest.useFakeTimers("modern").setSystemTime(new Date("2020-01-01").getTime());
  mock_pgQuery.mockReturnValue({ rows: [] });

  mock_matomoApi.mockImplementation((options, cb) => {
    return cb(null, [
      {
        ...matomoVisit,
        idVisit: 123,
      },
    ]);
  });

  await run();

  expect(mock_matomoApi.mock.calls.length).toEqual(170);

  // const expectedMatomoCalls = [
  //   ["2019-07-12", "42"],
  //   ["2019-07-13", "42"],
  //   ["2019-07-14", "42"],
  // ];

  // // check matomo requests
  // expect(mock_matomoApi.mock.calls.length).toEqual(expectedMatomoCalls.length);
  // expect(mock_matomoApi.mock.calls.map((c) => [c[0].date, c[0].idSite])).toEqual(expectedMatomoCalls);
  // // check the 4 events inserted
  // expect(mock_pgQuery.mock.calls.length).toEqual(1 + expectedMatomoCalls.length * (2 + 1));
  // expect(mock_pgQuery.mock.calls).toMatchSnapshot();
});
