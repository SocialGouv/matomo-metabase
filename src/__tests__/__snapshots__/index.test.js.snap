// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run: should create table 1`] = `
Array [
  "CREATE TABLE IF NOT EXISTS matomo
  (
    idsite                      text,
    idvisit                     text,
    actions                     text,
    country                     text,
    region                      text,
    city                        text,
    operatingsystemname         text,
    devicemodel                 text,
    devicebrand                 text,
    visitduration               text,
    dayssincefirstvisit         text,
    visitortype                 text,
    sitename                    text,
    userid                      text,
    serverdateprettyfirstaction date,
    action_id                   text UNIQUE,
    action_type                 text,
    action_eventcategory        text,
    action_eventaction          text,
    action_eventname            text,
    action_eventvalue           text,
    action_timespent            text,
    action_timestamp            timestamp with time zone,
    usercustomproperties        json
)",
  Array [],
]
`;

exports[`run: should use today date if nothing in DB 1`] = `
Array [
  Array [
    "CREATE TABLE IF NOT EXISTS matomo
  (
    idsite                      text,
    idvisit                     text,
    actions                     text,
    country                     text,
    region                      text,
    city                        text,
    operatingsystemname         text,
    devicemodel                 text,
    devicebrand                 text,
    visitduration               text,
    dayssincefirstvisit         text,
    visitortype                 text,
    sitename                    text,
    userid                      text,
    serverdateprettyfirstaction date,
    action_id                   text UNIQUE,
    action_type                 text,
    action_eventcategory        text,
    action_eventaction          text,
    action_eventname            text,
    action_eventvalue           text,
    action_timespent            text,
    action_timestamp            timestamp with time zone,
    usercustomproperties        json
)",
    Array [],
  ],
  Array [
    "select action_timestamp from matomo order by action_timestamp desc limit 1",
  ],
  Array [
    "SELECT COUNT(distinct idvisit) FROM matomo WHERE action_timestamp::date='2021-09-02';",
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, usercustomproperties)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      123,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "123_0",
      "event",
      "Ecommerce",
      "Cart change",
      "added - Basic Wetsuit",
      1,
      48,
      "2021-08-20T21:35:18.000Z",
      Object {
        "page-author": "Julien",
        "post-age": "-430 days",
      },
    ],
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, usercustomproperties)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      123,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "123_1",
      "action",
      undefined,
      undefined,
      undefined,
      undefined,
      "2",
      "2021-08-20T21:30:25.000Z",
      Object {},
    ],
  ],
]
`;
